File: win/directx/sdk-9.0/Samples/VB.Net/AudioVideo/Texture/Texture.vb
'-----------------------------------------------------------------------------
' File: texture.cs
'
' Copyright (c) 2000-2002 Microsoft Corporation. All rights reserved.
'-----------------------------------------------------------------------------
Imports System
Imports System.Drawing
Imports System.Windows.Forms
Imports Microsoft.DirectX
Imports Microsoft.DirectX.Direct3D
Imports Microsoft.DirectX.AudioVideoPlayback
Imports Direct3D = Microsoft.DirectX.Direct3D


Namespace VideoTexture
    _
    Public Class VideoTexture
        Inherits GraphicsSample
        ' Our global variables for this project
        Private vertexBuffer As vertexBuffer = Nothing
        Private texture As texture = Nothing
        Private videoTexture As Video = Nothing



        Public Sub New()
            ' Set it's caption
            Me.Text = "Direct3D Video Textures"
            isUsingMenus = False
            isMultiThreaded = False
        End Sub 'New


        Sub RenderIt(ByVal sender As Object, ByVal e As TextureRenderEventArgs)
            'Set our texture so we can render it
            texture = e.Texture
            RenderTexture()
        End Sub

        Sub MovieOver(ByVal sender As Object, ByVal e As EventArgs)
            videoTexture.Stop()
            videoTexture.Play()
        End Sub 'MovieOver

        Public Sub OnCreateVertexBuffer(ByVal sender As Object, ByVal e As EventArgs)
            Dim vb As VertexBuffer = CType(sender, VertexBuffer)
            ' Create a vertex buffer (100 customervertex)
            Dim verts As CustomVertex.PositionNormalTextured() = CType(vb.Lock(0, 0), CustomVertex.PositionNormalTextured()) ' Lock the buffer (which will return our structs)
            Dim i As Integer
            For i = 0 To 49
                ' Fill up our structs
                Dim theta As Single = CSng(2 * Math.PI * i) / 49
                verts((2 * i)).SetPosition(New Vector3(CSng(Math.Sin(theta)), -1, CSng(Math.Cos(theta))))
                verts((2 * i)).SetNormal(New Vector3(CSng(Math.Sin(theta)), 0, CSng(Math.Cos(theta))))
                verts((2 * i)).Tu = CSng(i) / (50 - 1)
                verts((2 * i)).Tv = 1.0F
                verts((2 * i + 1)).SetPosition(New Vector3(CSng(Math.Sin(theta)), 1, CSng(Math.Cos(theta))))
                verts((2 * i + 1)).SetNormal(New Vector3(CSng(Math.Sin(theta)), 0, CSng(Math.Cos(theta))))
                verts((2 * i + 1)).Tu = CSng(i) / (50 - 1)
                verts((2 * i + 1)).Tv = 0.0F
            Next i
            ' Unlock (and copy) the data
            vb.Unlock()
        End Sub 'OnCreateVertexBuffer

        Protected Overrides Sub RestoreDeviceObjects(ByVal sender As Object, ByVal e As System.EventArgs)

            If vertexBuffer Is Nothing Then
                ' Now Create the VB
                vertexBuffer = New VertexBuffer(GetType(CustomVertex.PositionNormalTextured), 100, Device, Usage.WriteOnly, CustomVertex.PositionNormalTextured.Format, Pool.Default)
                AddHandler vertexBuffer.Created, AddressOf Me.OnCreateVertexBuffer
                Me.OnCreateVertexBuffer(vertexBuffer, Nothing)
            ElseIf vertexBuffer.Disposed Then
                ' Now Create the VB
                vertexBuffer = New VertexBuffer(GetType(CustomVertex.PositionNormalTextured), 100, Device, Usage.WriteOnly, CustomVertex.PositionNormalTextured.Format, Pool.Default)
                AddHandler vertexBuffer.Created, AddressOf Me.OnCreateVertexBuffer
                Me.OnCreateVertexBuffer(vertexBuffer, Nothing)
            End If

            Device.RenderState.Ambient = System.Drawing.Color.White
            ' Turn off culling, so we see the front and back of the triangle
            Device.RenderState.CullMode = Cull.None
            ' Turn off D3D lighting
            Device.RenderState.Lighting = False
            ' Turn on the ZBuffer
            Device.RenderState.ZBufferEnable = True

            Device.SamplerState(0).AddressU = TextureAddress.Clamp
            Device.SamplerState(0).AddressV = TextureAddress.Clamp

            Try

                Dim path As String = DXUtil.FindMediaFile(Nothing, "ruby.avi")
                videoTexture = Video.FromFile(path)
                AddHandler videoTexture.Ending, AddressOf Me.MovieOver
                AddHandler videoTexture.TextureReadyToRender, AddressOf Me.RenderIt

            ' Now start rendering to our texture
                videoTexture.RenderToTexture(Device)
            Catch err As Exception
                MessageBox.Show(String.Format("An error has occurred that will not allow this sample to continue.\r\nException={0}", err.ToString()), "This sample must exit.", MessageBoxButtons.OK, MessageBoxIcon.Information)
                Me.Close()
                Throw err
            End Try

        End Sub

        Private Sub SetupMatrices()
            ' For our world matrix, we will just rotate the object about the y-axis.
            Device.Transform.World = Matrix.RotationAxis(New Vector3(CSng(Math.Cos((Environment.TickCount / 250.0F))), 1, CSng(Math.Sin((Environment.TickCount / 250.0F)))), Environment.TickCount / 1000.0F)

            ' Set up our view matrix. A view matrix can be defined given an eye point,
            ' a point to lookat, and a direction for which way is up. Here, we set the
            ' eye five units back along the z-axis and up three units, look at the
            ' origin, and define "up" to be in the y-direction.
            Device.Transform.View = Matrix.LookAtLH(New Vector3(0.0F, 3.0F, -5.0F), New Vector3(0.0F, 0.0F, 0.0F), New Vector3(0.0F, 1.0F, 0.0F))

            ' For the projection matrix, we set up a perspective transform (which
            ' transforms geometry from 3D view space to 2D viewport space, with
            ' a perspective divide making objects smaller in the distance). To build
            ' a perpsective transform, we need the field of view (1/4 pi is common),
            ' the aspect ratio, and the near and far clipping planes (which define at
            ' what distances geometry should be no longer be rendered).
            Device.Transform.Projection = Matrix.PerspectiveFovLH(CSng(Math.PI) / 4.0F, 1.0F, 1.0F, 100.0F)
        End Sub 'SetupMatrices


        Private Sub RenderTexture()
            'Clear the backbuffer to a blue color 
            Device.Clear(ClearFlags.Target Or ClearFlags.ZBuffer, System.Drawing.Color.Blue, 1.0F, 0)

            'Begin the scene
            Device.BeginScene()
            ' Setup the world, view, and projection matrices
            SetupMatrices()
            ' Setup our texture. Using textures introduces the texture stage states,
            ' which govern how textures get blended together (in the case of multiple
            ' textures) and lighting information. In this case, we are modulating
            ' (blending) our texture with the diffuse color of the vertices.
            Device.SetTexture(0, texture)
            Device.SetStreamSource(0, vertexBuffer, 0)
            Device.VertexFormat = CustomVertex.PositionNormalTextured.Format
            Device.DrawPrimitives(PrimitiveType.TriangleStrip, 0, 4 * 25 - 2)
            'End the scene
            Device.EndScene()
            ' Update the screen
            Device.Present()
        End Sub 'Render


        '/ <summary>
        '/ The main entry point for the application.
        '/ </summary>
        Shared Sub Main()

            Dim frm As New VideoTexture()
            Try
                If Not frm.CreateGraphicsSample() Then ' Initialize Direct3D
                    MessageBox.Show("Could not initialize Direct3D.  This tutorial will exit.")
                    frm.Dispose()
                    Return
                End If

                If (frm.IsHandleCreated) Then
                    frm.ShowDialog() ' Only need to show the dialog if we haven't closed
                End If
            Finally
                frm.Dispose()
            End Try
        End Sub 'Main

    End Class 'VideoTexture 
End Namespace 'VideoTexture 

File: win/directx/sdk-9.0/Samples/C#/AudioVideo/Texture/texture.cs
//-----------------------------------------------------------------------------
// File: texture.cs
//
// Copyright (c) 2000-2002 Microsoft Corporation. All rights reserved.
//-----------------------------------------------------------------------------
using System;
using System.Drawing;
using System.Windows.Forms;
using Microsoft.DirectX;
using Microsoft.DirectX.Direct3D;
using Microsoft.DirectX.AudioVideoPlayback;
using Direct3D=Microsoft.DirectX.Direct3D;

namespace VideoTexture
{
	public class VideoTexture : GraphicsSample
	{
		VertexBuffer vertexBuffer = null;
		Texture texture = null;
        Video videoTexture = null;


		public VideoTexture()
		{
			// Set it's caption
			this.Text = "Direct3D Video Textures";
            isUsingMenus = false;
            isMultiThreaded = true;
		}

        void RenderIt(object sender, TextureRenderEventArgs e)
        {
            lock(this)
            {
                // Set our texture so we can render it
                texture = e.Texture;
                RenderTexture();
            }
        }
        void MovieOver(object sender, EventArgs e)
        {
            videoTexture.Stop();
            videoTexture.Play();
        }
		public void OnCreateVertexBuffer(object sender, EventArgs e)
		{
            VertexBuffer vb = (VertexBuffer)sender;
			// Create a vertex buffer (100 customervertex)
			CustomVertex.PositionNormalTextured[] verts = (CustomVertex.PositionNormalTextured[])vb.Lock(0,0); // Lock the buffer (which will return our structs)
			for (int i = 0; i < 50; i++)
			{
				// Fill up our structs
				float theta = (float)(2 * Math.PI * i) / 49;
				verts[2 * i].SetPosition(new Vector3((float)Math.Sin(theta), -1, (float)Math.Cos(theta)));
				verts[2 * i].SetNormal(new Vector3((float)Math.Sin(theta), 0, (float)Math.Cos(theta)));
				verts[2 * i].Tu       = ((float)i)/(50-1);
				verts[2 * i].Tv       = 1.0f;
				verts[2 * i + 1].SetPosition(new Vector3((float)Math.Sin(theta), 1, (float)Math.Cos(theta)));
				verts[2 * i + 1].SetNormal(new Vector3((float)Math.Sin(theta), 0, (float)Math.Cos(theta)));
				verts[2 * i + 1].Tu       = ((float)i)/(50-1);
				verts[2 * i + 1].Tv       = 0.0f;
			}
			// Unlock (and copy) the data
			vb.Unlock();
		}
		private void SetupMatrices()
		{
			// For our world matrix, we will just rotate the object about the y-axis.
			device.Transform.World = Matrix.RotationAxis(new Vector3((float)Math.Cos(Environment.TickCount / 250.0f),1,(float)Math.Sin(Environment.TickCount / 250.0f)), Environment.TickCount / 1000.0f );

			// Set up our view matrix. A view matrix can be defined given an eye point,
			// a point to lookat, and a direction for which way is up. Here, we set the
			// eye five units back along the z-axis and up three units, look at the
			// origin, and define "up" to be in the y-direction.
			device.Transform.View = Matrix.LookAtLH( new Vector3( 0.0f, 3.0f,-5.0f ), new Vector3( 0.0f, 0.0f, 0.0f ), new Vector3( 0.0f, 1.0f, 0.0f ) );

			// For the projection matrix, we set up a perspective transform (which
			// transforms geometry from 3D view space to 2D viewport space, with
			// a perspective divide making objects smaller in the distance). To build
			// a perpsective transform, we need the field of view (1/4 pi is common),
			// the aspect ratio, and the near and far clipping planes (which define at
			// what distances geometry should be no longer be rendered).
			device.Transform.Projection = Matrix.PerspectiveFovLH( (float)Math.PI / 4.0f, 1.0f, 1.0f, 100.0f );
		}

        


        protected override void RestoreDeviceObjects(object sender, System.EventArgs e)
        {
            // Now Create the VB
            if ((vertexBuffer == null) || (vertexBuffer.Disposed))
            {
                vertexBuffer = new VertexBuffer(typeof(CustomVertex.PositionNormalTextured), 100, device, Usage.WriteOnly, CustomVertex.PositionNormalTextured.Format, Pool.Default);
                vertexBuffer.Created += new System.EventHandler(this.OnCreateVertexBuffer);
                this.OnCreateVertexBuffer(vertexBuffer, null);
            }

            device.RenderState.Ambient = System.Drawing.Color.White;
            // Turn off culling, so we see the front and back of the triangle
            device.RenderState.CullMode = Cull.None;
            // Turn off D3D lighting
            device.RenderState.Lighting = false;
            // Turn on the ZBuffer
            device.RenderState.ZBufferEnable = true;

            device.SamplerState[0].AddressU = TextureAddress.Clamp;
            device.SamplerState[0].AddressV = TextureAddress.Clamp;

            string path = DXUtil.FindMediaFile(null, "ruby.avi");

            try
            {
                videoTexture = Video.FromFile(path);
                videoTexture.Ending += new System.EventHandler(this.MovieOver);
                videoTexture.TextureReadyToRender += new TextureRenderEventHandler(this.RenderIt);

                // Now start rendering to our texture
                videoTexture.RenderToTexture(device);
            }
            catch(Exception err)
            {
                MessageBox.Show(string.Format("An error has occurred that will not allow this sample to continue.\r\nException={0}", err.ToString()), "This sample must exit.", MessageBoxButtons.OK, MessageBoxIcon.Information);
                this.Close();
                throw err;
            }
        }

		private void RenderTexture()
		{
            //Clear the backbuffer to a blue color 
            device.Clear(ClearFlags.Target | ClearFlags.ZBuffer, System.Drawing.Color.Blue, 1.0f, 0);
            //Begin the scene
            device.BeginScene();
            // Setup the world, view, and projection matrices
            SetupMatrices();
            // Setup our texture. Using textures introduces the texture stage states,
            // which govern how textures get blended together (in the case of multiple
            // textures) and lighting information. In this case, we are modulating
            // (blending) our texture with the diffuse color of the vertices.
            device.SetTexture(0,texture);
            device.SetStreamSource(0, vertexBuffer, 0);
            device.VertexFormat = CustomVertex.PositionNormalTextured.Format;
            device.DrawPrimitives(PrimitiveType.TriangleStrip, 0, (4*25)-2);
            //End the scene
            device.EndScene();
            device.Present();
        }

		/// <summary>
		/// The main entry point for the application.
		/// </summary>
		static void Main() 
		{
            using (VideoTexture frm = new VideoTexture())
            {
                if (!frm.CreateGraphicsSample()) // Initialize Direct3D
                {
                    MessageBox.Show("Could not initialize Direct3D.  This sample will exit.");
                    return;
                }

                if (frm.IsHandleCreated) // Only run if we haven't closed the form yet (ie, from an error)
                    frm.ShowDialog();
            }
		}
	}
}
